docs/design/storage-node-automatic-updater.md

```diff
@@ -0,0 +1,143 @@
+
+The Updater has several responsibilities:
+
+1. Figure out whether something needs to be updated with gradual rollout/
+1. Safely download the binaries.
+1. Safely update the binaries.
+1. Safely restart the binaries.
+
+## Design
+
+The Updater has several steps it takes, contact version server, download, update, restart.
+
+### Checking for updates
+
+Update check will regularly, with jitter, contact Version Server, which responds with a message:
+
+```json
+{
+  "processes": {
+    "storagenode": {
+      "allowed_version_minimum": "0.3.4",
+      "suggested_version": "0.5.1",
+      "rollout": {
+        "active": true,
+        "target_version": "0.5.2",
+        "rollout_seed": "04123bacde",
+        "rollout_cursor": "40"
+      }
+    }
+  }
+}
+```
+
+When there is a newer version is available it needs to calculate whether it needs to update. To check whether rollout has reached this node it needs to calculate `hash(rollout_seed, node_id) < rollout_cursor`. This exact behavior may differ for canary nodes, which always get the latest version.
+
+* The update check must verify that it is a trusted server.
+* The update check should have a jitter to avoid a stampeding herd. See http://highscalability.com/blog/2012/4/17/youtube-strategy-adding-jitter-isnt-a-bug.html for more information.
+
+### Canary releases
+
+Canary nodes are the first storage nodes that will receive a new release. They will receive it even before first nodes in the gradual rollout. Canary nodes are on the front line on the risk of receiving a bad new release. Only after confirming that canary nodes are behaving correctly after the update, the gradual rollout will begin.
+
+It shall be possible for Storage Node Operator to opt-in to the canary release channel. We expect that those will be mostly Storj employees and enthusiastic community members.
+
+Canary nodes will query the Version Server on a different designated URL to find the latest canary version.
+
+### Downloading the binaries
+
+Once we have decided on a new version we need to download the new version. We will download the appropriate release from a trusted server beside the current binary (instead of temporary directory).
+
+Once we have successfully downloaded we must verify that the binary signature is valid.
+
+Possible problems:
+* Downloading could fail.
+* Out-of-space for downloading.
+* Filesystem is read-only.
+* Corrupted binary.
+* Man in the Middle attacks.
+    * Verify binary hashes and the binary signature.
+* Downloaded file may be quarantined by the anti-virus or blocked by the firewall.
+
+### Updating the binaries
+
+To update the binaries we can take two approaches. 
+
+1. Rename `storagenode.exe` into `storagenode.old.<release>.exe`.
+1. Rename `storagenode.<release>.exe` into `storagenode.exe`.
+1. Restart the service using Windows API.
+1. Delete `storagenode.old.<release>.exe`.
```

Comment:
Maybe we should give users possibility to revert update to last working version and not delete `storagenode.old.*`?

Reply:
oh ok, I just read below `We won't support rolling back to the previous version.` :)